PP-480: Added more regressions for job equivalence classes

anamikau opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-480](https://pbspro.atlassian.net/browse/PP-480)**

#### Problem description
* Added more regression tests

#### Cause / Analysis
* Added more regression tests

#### Solution description
* Added more regression tests

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

hirenvadalia commented:
@anamikau Can you change title to something specific instead "Added more regressions"? Also in commit message 'pp-480' should be 'PP-480' and ':' missing in PR title.

anamikau commented:
I forgot to run pep8 hence travis failed

bhroam commented on test/tests/performance/pbs_equiv_classes_perf.py:
```diff
@@ -108,12 +112,159 @@ def test_basic(self):
+        """
+        Test the performance with hard and soft limits
+        on resources
+        """
```
I don't see this test as a performance test.  To see performance you will need to show the scheduler being slow and then it being faster using the equiv class feature.  Right now you submit 8 different classes and run a cycle.  You then run another cycle without changing anything.  These two cycles should take the same amount of time.  I see two ways you can go with this test.  You could do something similar to what test_basic() does.  Submit N different classes and run one cycle.  You can then break those jobs up into your 8 classes and rerun your cycle.  Show the difference in time.  The second thing you could do is to move this to the functional tests.  It is useful there because there isn't really a test that tests all the different ways to differentiate classes.  The tests I wrote basically would test one of them at a time.  This one tests 8 different ones.

bhroam commented on test/tests/performance/pbs_equiv_classes_perf.py:
```diff
@@ -108,12 +112,159 @@ def test_basic(self):
+        a = {'preempt_prio':
+             'express_queue, normal_jobs, server_softlimits,\
+             queue_softlimits,server_softlimits+queue_softlimits'}
+        self.scheduler.set_sched_config(a)
```
You don't need to set preempt_prio to submit jobs with soft limits.  Setting preempt_prio just allows actual preemption to take place.  I don't see you doing that here.  Now testing this feature with preemption isn't necessarily a bad idea.  A suspended job uses a special select spec to make sure it is placed back on the same nodes it was on before.  If a suspended job is not in a class by itself, it could cause problems.  I don't remember if I took care of this case.  Maybe write a functional test for this?

bhroam commented on test/tests/performance/pbs_equiv_classes_perf.py:
```diff
@@ -108,12 +112,159 @@ def test_basic(self):
+
+        self.logger.info('Cycle 1: %d Cycle 2: %d' %
+                         (cycle1_time, cycle2_time))
+        self.assertTrue(cycle1_time >= cycle2_time)
```
Maybe use self.assertGreaterEqual()?

anamikau commented on test/tests/functional/pbs_equiv_classes.py:
```diff
@@ -1050,3 +1051,353 @@ def test_ded_time_queue(self):
+
+        # One equivalence class as execjob_begin hook runs
+        # after job is scheduled by scheduler
+        m = self.scheduler.log_match("Number of job equivalence classes: 3",
```
I was expecting 1 but I observed 3 being reported.

bhroam replied:
You get 3 based on your select specs.  Each is requesting an individual vnode, so they all differ from each other.

anamikau replied:
Actually the reason I was expecting one because of the execjob_begin hook. I thought it will be picked up in the next cycle but I was wrong. anyways I have corrected it now.

anamikau commented:
@bhroam, I have addressed your review comments. Also moved the previously written performance test to functional and modified the performance test. I will attach the results to JIRA.

bhroam commented on test/tests/functional/pbs_equiv_classes.py:
```diff
@@ -1181,10 +1151,9 @@ def test_fairshare(self):
-        self.assertTrue(m)
+        self.scheduler.log_match("Number of job equivalence classes: 1",
+                                 max_attempts=10,
+                                 starttime=int(self.t) - 1)
```
Why the -1?  This can be dangerous depending on the length of tests.  You might match a previous cycle's log message.  Also, some times you int(self.t) in the log_match() and other times it is int()'d at assignment time.  I'd like to see this consistent.

bhroam commented on test/tests/functional/pbs_equiv_classes.py:
```diff
@@ -1392,12 +1353,133 @@ def test_incr_decr(self):
+             'group_list': TSTGRP1, ATTR_q: 'workq2'}
+        J = Job(TEST_USER1, attrs=a)
+        for n in range(10):
+            self.server.submit(J)
```
Use the submit_jobs() method instead of directly submitting jobs (and for the rest of the test)

bhroam commented on test/tests/performance/pbs_equiv_classes_perf.py:
```diff
@@ -51,19 +51,41 @@ def setUp(self):
+        self.server.create_vnodes('vnode', a, 10000, self.mom,
+                                  sharednode=False)
+
+    def get_cycle_time(self):
```
Maybe change the name of this function to run_and_get_cycle_time()?

bhroam commented on test/tests/performance/pbs_equiv_classes_perf.py:
```diff
@@ -51,19 +51,41 @@ def setUp(self):
+
+    def get_cycle_time(self):
+        """
+        Calculate the time between scheduling cycles
```
You aren't calculating the time between cycles, you are running a cycle and then calculating the time of that cycle.

bhroam commented on test/tests/performance/pbs_equiv_classes_perf.py:
```diff
@@ -51,19 +51,41 @@ def setUp(self):
-        a = {'resources_available.ncpus': 1, 'resources_available.mem': '8gb'}
-        self.server.create_vnodes('vnode', a, 100, self.mom,
-                                  sharednode=False)
-
```
Why remove the creation of the vnodes?

anamikau replied:
They are not removed but moved. I moved it back to the setUp.

bhroam commented on test/tests/performance/pbs_equiv_classes_perf.py:
```diff
@@ -177,94 +160,109 @@ def test_server_queue_limit(self):
+                         600 classes is %d,\
+                         700 classes is %d,\
+                         800 classes is %d"
+                         % (cyc1, cyc2, cyc3, cyc4, cyc5, cyc6, cyc7, cyc8))
```
With the string continuation with the \\, won't the white space be taken as literal white space?  I think you want to use multiple quoted strings.  You can use "a"     "b" == "ab".  Also, some of your jobs may run.  By the time you reach the bottom of your 80,000 job test, you won't have all 80,000 jobs any more.  Some will have completed.  In test_basic() I made sure the jobs couldn't run.  You may want to do that here.

bhroam commented on test/tests/functional/pbs_equiv_classes.py:
```diff
@@ -1392,12 +1353,133 @@ def test_incr_decr(self):
+        a = {'preempt_prio':
+             'express_queue, normal_jobs, server_softlimits,\
+             queue_softlimits,server_softlimits+queue_softlimits'}
+        self.scheduler.set_sched_config(a)
```
You don't need to set this unless you really want to.  The only reason preempt_prio needs to be set is if preemption needs to happen.  I don't believe you want any preemption to actually happen during this test.

bhroam commented on test/tests/functional/pbs_equiv_classes.py:
```diff
@@ -1392,12 +1353,133 @@ def test_incr_decr(self):
+        self.server.manager(
+            MGR_CMD_SET, NODE,
+            {'resources_available.ncpus': '50'},
+            id=self.mom.shortname)
```
Why 50 cpus?  It will just let more jobs run.  I'm not sure you want that.

bhroam commented on test/tests/performance/pbs_equiv_classes_perf.py:
```diff
@@ -56,12 +56,13 @@ def setUp(self):
         """
-        Calculate the time between scheduling cycles
+        Running the scheudling cycle and Calculating
+        the time between them
```
How about "Run a scheduling cycle and calculate its duration" ?

anamikau replied:
ok. done.

anamikau commented:
@bhroam and @bayucan, can you please look at minor changes to this PR?

bhroam commented:
@anamikau, everything looks good.  I sign off.  Once @bayucan signs off again, please squash and rebase and I'll merge it in.

anamikau commented:
@bhroam, rebased and squashed my commits. please take it forward.

bhroam commented:
I merged in the pull request
